WooCommerce/Classes/Tools/Logging/WooCrashLoggingStack.swift

```diff
@@ -0,0 +1,145 @@
+        do {
+            crashLoggingDataProvider.appIsCrashing = true
+            try crashLogging.logErrorAndWait(error, userInfo: userInfo, level: sentrySeverity(with: level))
+        } catch {
+            DDLogError("⛔️ Unable to send startup error message to Sentry: \(error)")
+        }
+        fatalError(error.localizedDescription)
+    }
+
+    func setNeedsDataRefresh() {
+        crashLogging.setNeedsDataRefresh()
+    }
+
+    var queuedLogFileCount: Int {
+        eventLogging.queuedLogFiles.count
+    }
+
+    private func willEnterForeground(note: Foundation.Notification) {
+        self.eventLogging.uploadNextLogFileIfNeeded()
+        DDLogDebug("📜 Resumed encrypted log upload queue due to app entering foreground")
+    }
+
+    private func sentrySeverity(with storageSeverity: SeverityLevel) -> SentryLevel {
+        switch storageSeverity {
+        case .fatal:
+            return .fatal
+        case .error:
+            return .error
+        case .warning:
+            return .warning
+        case .info:
+            return .info
+        case .debug:
+            return .debug
+        }
+    }
+}
+
+class WCCrashLoggingDataProvider: CrashLoggingDataProvider {
+    /// Indicates that app is in an inconsistent state and we don't want to start asking it for metadata
+    fileprivate var appIsCrashing = false
+
+    init() {
+        NotificationCenter.default.addObserver(self, selector: #selector(updateCrashLoggingSystem(_:)), name: .defaultAccountWasUpdated, object: nil)
+        NotificationCenter.default.addObserver(self, selector: #selector(updateCrashLoggingSystem(_:)), name: .logOutEventReceived, object: nil)
+        NotificationCenter.default.addObserver(self, selector: #selector(updateCrashLoggingSystem(_:)), name: .StoresManagerDidUpdateDefaultSite, object: nil)
+    }
+
+    var userHasOptedOut: Bool {
+        return !CrashLoggingSettings.didOptIn
+    }
+
+    var currentUser: TracksUser? {
+        guard !appIsCrashing else {
+            return nil
+        }
+
+        guard let account = ServiceLocator.stores.sessionManager.defaultAccount else {
+            return nil
+        }
+
+        return TracksUser(userID: "\(account.userID)", email: account.email, username: account.username)
+    }
+
+    var sentryDSN: String {
+        return ApiCredentials.sentryDSN
+    }
+
+    var buildType: String {
+        return BuildConfiguration.current.rawValue
+    }
+
+    @objc func updateCrashLoggingSystem(_ notification: Notification) {
+        /// Bumping this call to a later run loop is a little bit hack-y, but because the `StoresManager` fires the events
+        /// we're interested as part of its initialization, we need to wait for that initalization to be complete before
+        /// taking action – otherwise the application will deadlock.
+        DispatchQueue.main.async {
+            ServiceLocator.crashLogging.setNeedsDataRefresh()
+        }
```

Comment:
What about if the app is about to crash in the current iteration of the runloop (and will thus never execute the dispatched code), is that a risk/problem?

Reply:
This was already there and I only tried to adapt the current logic to the new API: https://github.com/woocommerce/woocommerce-ios/pull/4443/files#diff-e97506f77d6983646e4da3f634625fce913b97e6e19cf430a8248ab5ed23ac62

I don't understand all the reasoning behind this, but it seems to me that the worst case scenario here is that a crash would have the wrong (to an extent) username attached to it.